server/src/main/java/org/elasticsearch/plugins/FieldPredicate.java

```diff
@@ -88,4 +88,36 @@ public String toString() {
             return first + " then " + second;
         }
     }
+
+    class Or implements FieldPredicate {
+        private static final long SHALLOW_RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(And.class);
+
+        private final FieldPredicate first;
+        private final FieldPredicate second;
+
+        public Or(FieldPredicate first, FieldPredicate second) {
+            this.first = first;
+            this.second = second;
+        }
+
+        @Override
+        public boolean test(String field) {
+            return first.test(field) || second.test(field);
+        }
+
+        @Override
+        public String modifyHash(String hash) {
+            return second.modifyHash(first.modifyHash(hash));
```

Comment:
This is the same as for `And` -- is that a problem, and if so should I add a prefix or something like that to distinguish?

Reply:
Looking at this, it does seem like it could potentially cause hash collision (since hash is used as a key in a map). So probably prefixing it would be safer.